nifty/pipeline/nifsUtils.py

```diff
@@ -1163,3 +1185,66 @@ def MEFarith(MEF, image, op, result):
             iraf.imarith(operand1=result+'['+str(i)+']', op=op, operand2 = image, result = result+'['+str(i)+', overwrite]', divzero = 0.0)
 
 #-----------------------------------------------------------------------------#
+
+def downloadQueryCadc(program, directory='./rawData'):
+    """
+    Finds and downloads all CADC files for a particular gemini program ID to
+    the current working directory.
+    """
+
+    cadc = Cadc()
+    job = cadc.create_async("SELECT observationID, publisherID, productID FROM caom2.Observation \
+                             AS o JOIN caom2.Plane AS p ON o.obsID=p.obsID \
+                             WHERE instrument_name='NIFS' AND proposal_id={}".format("'"+program+"'"))
+    job.run().wait()
+    job.raise_if_error()
+    result = job.fetch_result().to_table()
+
+    # Store product id's for later
+    pids = list(result['productID'])
+
+    urls = cadc.get_data_urls(result)
+    for url, pid in zip(urls, pids):
+        try:
+            filename = getFile(url)
+            shutil.move(filename, directory+'/'+filename)
+            logging.debug("Downloaded {}".format(filename))
+        except Exception as e:
+            logging.error("A frame failed to download.")
+            raise e
+
+
+def getFile(url):
+    """
+    Gets a file from the specified url and returns the filename.
+    """
+    r = requests.get(url, stream=True)
+    # Parse out filename from header
+    try:
+        filename = re.findall("filename=(.+)", r.headers['Content-Disposition'])[0]
+    except KeyError:
+        # 'Content-Disposition' header wasn't found, so parse filename from URL
+        # Typical URL looks like:
+        # https://www.cadc-ccda.hia-iha.nrc-cnrc.gc.ca/data/pub/GEM/N20140505S0114.fits?RUNID=mf731ukqsipqpdgk
+        filename = (url.split('/')[-1]).split('?')[0]
+    # Write the fits file
+    with open(filename, 'wb') as f:
```

Comment:
I would probably think writing to a temporary directory or an application directory. You could expect some failures while downloading the file potentially leaving incomplete files behind. Or write them straight into the destination directory with a temporary name (starting with '.' for example). This is where you'd want to compute the md5 of the received data and compare it with the `content-MD5` of the file (assuming the GEMINI also supports it).

Reply:
Very good idea. I went with writing to a '.temp-<filename>' file in the working directory for now and doing the md5 check there. I could change it to write the temp file in the download directory as well. Changes are in ff758ba and cf023d9.

Reply:
Is there a mechanism to get rid of the failed temporary files? Are they deleted when errors occur? Otherwise they might pile up consuming user's quota. You could delete the old ones on startup or tear down. Might want to do that only for older files if multiple instances of the script are executed at the same time. Alternatively, you could use something like the `tempfile` package to store them in a temporary directory. The cleanup is someone else's job in that case.
'.temp-' is used in multiple places and should be a const. To avoid misspelling bugs.

Reply:
I recommend using `tempfile.TemporaryFile(mode='w+b', buffering=None, encoding=None, newline=None, suffix=None, prefix=None, dir=None, *, errors=None)`

Set the dir=$CACHEDIR and prefix=filename 

On success mv the file to the correct filename ... if fail raise a exception.  file cleanup is handled by python.